File: python/questions/RandomizedSet.py
import random


class RandomizedSet:

    def __init__(self):
        self.dt = {}
        self.v = []

    def insert(self, val: int) -> bool:
        idx = self.dt.get(val, -1)
        if idx == -1:
            self.dt[val] = len(self.v)
            self.v.append(val)
            return True
        return False

    def remove(self, val: int) -> bool:
        idx = self.dt.get(val, -1)
        if idx == -1:
            return False
        self.v[idx] = self.v[-1]
        self.dt[self.v[-1]] = idx
        self.dt.pop(val)
        self.v.pop()
        return True

    def getRandom(self) -> int:
        return self.v[random.randrange(len(self.v))]

File: typescript/questions/RandomizedSet.ts
class RandomizedSet {
  private dt = new Map<number, number>();
  private v: number[] = [];

  constructor() {}

  insert(val: number): boolean {
    if (this.dt.has(val)) {
      return false;
    }
    this.dt.set(val, this.v.length);
    this.v.push(val);
    return true;
  }

  remove(val: number): boolean {
    const idx = this.dt.get(val);
    if (idx == undefined) {
      return false;
    }
    const ev = this.v[this.v.length - 1];
    this.dt.set(ev, idx);
    this.v[idx] = ev;
    this.v.pop();
    this.dt.delete(val);
    return true;
  }

  getRandom(): number {
    const idx = Math.floor(Math.random() * this.v.length);
    return this.v[idx];
  }
}
